tokio/src/sync/semaphore.rs

```diff
@@ -333,6 +333,76 @@ use std::sync::Arc;
 /// }
 /// ```
 ///
+/// ## Countdown latch
+///
+/// A `Semaphore` may be used to implement a countdown latch, a synchronization
+/// primitive where one task waits until a set number of other tasks to have
+/// completed before continuing.
+///
+/// The task awaiting the countdown should call [`Semaphore::acquire_many`] to
+/// acquire the a number of permits equal to the number of tasks which hold the
+/// latch open. The tasks whose completion is awaited by the latch should call
+/// [`Semaphore::add_permit`] with a single permit to increment the countdown
+/// latch as they complete. For example:
+///
+/// ```rust
+/// use tokio::sync::Semaphore;
+/// use std::{future::Future, sync::Arc};
+///
+/// /// A token that holds a countdown latch open until it is dropped.
+/// #[derive(Clone)]
+/// pub struct Countdown(Arc<Semaphore>);
```

Comment:
This `derive(Clone)` seems a bit weird to me. It makes it difficult to make sure you have the right number of tokens. In your example, you also create one token more than you need, but the extra one in main is not dropped until after the wait completes.

Reply:
yeah, that's a good point. this could maybe have some kind of `fn try_clone -> Option<Self>` that only allows creating as many guards as there are permits, or something? i'm not sure if the additional complexity is worth including in the example or not...

Reply:
Maybe just return a vector of them?